src/lib.rs

```diff
@@ -0,0 +1,251 @@
+///
+///  The version field is automatically written and check by the `Encoder` and `Decoder`
+///  respectively.
+
+#[macro_use]
+extern crate serde_derive;
+
+mod decode;
+mod encode;
+
+pub use decode::Decoder;
+pub use encode::Encoder;
+
+/// The version number of the data structures.
+const SER_VERSION: usize = 0;
+
+pub type CrateHash = u64;
+pub type DefIndex = u32;
+pub type BasicBlockIndex = u32;
+
+/// A mirror of the compiler's notion of a "definition ID".
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct DefId {
+    crate_hash: CrateHash,
+    def_idx: DefIndex,
+}
+
+impl DefId {
+    pub fn new(crate_hash: CrateHash, def_idx: DefIndex) -> Self {
+        Self {
+            crate_hash,
+            def_idx,
+        }
+    }
+}
+
+/// A MIR.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct Mir {
+    def_id: DefId,
+    blocks: Vec<BasicBlock>,
+}
+
+impl Mir {
+    /// Create a new MIR.
+    pub fn new(def_id: DefId, blocks: Vec<BasicBlock>) -> Self {
+        Self { def_id, blocks }
+    }
+}
+
+/// A MIR block.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct BasicBlock {
+    stmts: Vec<Statement>,
+    term: Terminator,
+}
+
+impl BasicBlock {
+    /// Create a new MIR block.
+    pub fn new(stmts: Vec<Statement>, term: Terminator) -> Self {
+        Self { stmts, term }
+    }
+}
+
+/// A MIR statement.
+/// FIXME to be populated.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub enum Statement {
+    Nop,
+}
+
+/// A call target.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub enum CallOperand {
+    /// A statically known function identified by its DefId.
+    Fn(DefId),
+    /// An unknown or unhandled callable.
+    Unknown, // FIXME -- Find out what else. Closures jump to mind.
+}
+
+/// A MIR block terminator.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub enum Terminator {
+    Goto {
+        target_bb: BasicBlockIndex,
+    },
+    SwitchInt {
+        target_bbs: Vec<BasicBlockIndex>,
+    },
+    Resume,
+    Abort,
+    Return,
+    Unreachable,
+    Drop {
+        target_bb: BasicBlockIndex,
+        unwind_bb: Option<BasicBlockIndex>,
+    },
+    DropAndReplace {
+        target_bb: BasicBlockIndex,
+        unwind_bb: Option<BasicBlockIndex>,
+    },
+    Call {
+        operand: CallOperand,
+        cleanup_bb: Option<BasicBlockIndex>,
+    },
+    Assert {
+        target_bb: BasicBlockIndex,
+        cleanup_bb: Option<BasicBlockIndex>,
+    },
+    Yield {
+        resume_bb: BasicBlockIndex,
+        drop_bb: Option<BasicBlockIndex>,
+    },
+    GeneratorDrop,
+    FalseEdges {
+        real_target_bb: BasicBlockIndex,
+    },
+    FalseUnwind {
+        real_target_bb: BasicBlockIndex,
+    },
+}
+
+/// The top-level meta-data type.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
```

Comment:
Do we need all of these to be derived?

Reply:
* Serialize and Deserialize, certainly.
 * Eq and PartialEq simplify tests, and may be useful later on too.
 * Debug is useful.
 * Clone is also used in tests, since the serialiser takes ownership of what's being serialised, and we need to compare the original input to what comes back from the round trip.

Reply:
Let's delete the others that we don't need then: it just leads to bloat (and also makes me wonder what it mean to compare these things for equality!).

Reply:
I think you are suggesting we should remove `Eq` and `PartialEq`. This is going to make testing quite awkward, as I'll have to do a big old restructuring with multiple comparisons.

E.g. instead of comparing two metadata's using an `assert_eq!(expect_md, got_md)` I'll have to have a match statement for each variant, then in the case of a MIR,  destructure each basic block, and the statements within, manually comparing.

Reply:
Ah, sorry, I see.

Reply:
You've got me thinking though. What does a derived `Eq` for a reference do? Compare addresses? We need to be careful, as that might not be what we want.

As it happens, we have no references in our md at the moment.

Reply:
All comparisons involve references: `fn eq(&self, other: &Self)` in essence. So that isn't an issue. [If you mean "pointers", though, then yes, I think the pointer is compared and not the thing its pointed to.]

Reply:
I see.

I'd like to think we are sane enough to not serialise a pointer :P